### PR TITLE
evp_md_init_internal: Avoid reallocating algctx if digest unchanged

### LINKED ISSUE #16947: evp_md_init_internal: type checks are superfluous + optimizations lost compared to 1.1.1
In 3.0 in evp_md_init_internal, even if you specify type == NULL in the function call, you always inevitably get one in this block

```
    if (type != NULL) {
        ctx->reqdigest = type;
    } else {
        if (ctx->digest == NULL) {
            ERR_raise(ERR_LIB_EVP, EVP_R_NO_DIGEST_SET);
            return 0;
        }
        type = ctx->digest;
    }
```

There are many checks for type == NULL or type != NULL after that point which are unnecessary.

At the beginning of the function, it also forces the algctx of the provider to be free'd even if the algorithm is re-used wasting an opportunity for optimization if the algorithm is the same and comes from the same provider.

For example, this block was optional before and only done if you changed the digest type:

```
    if (ctx->digest != NULL && ctx->digest->ctx_size > 0) {
        OPENSSL_clear_free(ctx->md_data, ctx->digest->ctx_size);
        ctx->md_data = NULL;
    }
```

Now it's unconditionally executed leading to more wasted time and resources (see legacy path for much of the old 1.1.1 implementation).

There's also some ctx->engine checks which are not inside #if !defined(OPENSSL_NO_ENGINE).

### PR DESCRIPTION
**Pull request by t8m**
Fixes #16947

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

**Comment by kaduk** [comment]
How do we know that the algctx doesn't need to do some reinitialization that is lighter-weight than a full reallocation, in the reinit scenarios?

**Comment by t8m** [comment]
@kaduk The algctx is reinitialized by the call here: https://github.com/openssl/openssl/pull/18105/files#diff-df03d3f4aedaf8c344451553a2dededfb98fc9c871899a55e4e07cdb5d7a228eR294

Or maybe I do not fully understand your question?

**Comment by kaduk** [comment]
> @kaduk The algctx is reinitialized by the call here: https://github.com/openssl/openssl/pull/18105/files#diff-df03d3f4aedaf8c344451553a2dededfb98fc9c871899a55e4e07cdb5d7a228eR294
> 
> Or maybe I do not fully understand your question?

@t8m you understood my question; I just didn't go looking far enough for additional context in that direction in the web UI.  Thanks!

**Comment by t8m** [comment]
ping for the second review

File: crypto/evp/digest.c
Comment (kaduk) [inline crypto/evp/digest.c:251]: If we are doing an implicit fetch (what this block goes on to do), would that risk having the provider change from under us and thereby invalidate any existing algctx?  Or are we guaranteed that if there is an algctx there must already be an associated provider?
In particular, this block can change the value of `type`!  So the check at L262 is testing different things than the check at R241.

File: crypto/evp/digest.c
Comment (kaduk) [inline crypto/evp/digest.c:220]: Just to confirm: if we follow the `skip_to_init` case ~20 lines up, that requires us to have been using an engine previously, which means we were not using a provided hash before?

**Review by kaduk — COMMENTED** [review]
The mechanics of the change are okay, it's just the latent risks from potential mis-reuse that I'm worried about (now that we're not always freeing the algctx).  One such risk you already replied to, and now that I did a more careful review I want to check about the implicit-fetch scenario (see inline).

File: crypto/evp/digest.c
Comment (t8m) [reply crypto/evp/digest.c:251]: If there is an algctx set, the ctx->digest is already the fetched (non-implicit) digest (see https://github.com/openssl/openssl/pull/18105/files#diff-df03d3f4aedaf8c344451553a2dededfb98fc9c871899a55e4e07cdb5d7a228eR280). So if the type is an implicit digest it cannot be equal to ctx->digest in the if above and the algctx will be freed.

The removed block below was actually completely bogus and the code was not broken just because we always freed the algctx anyway.

Comment (kaduk) [reply crypto/evp/digest.c:251]: Okay, so you are saying that the only way `ctx->algctx` can be set is through the "R280" line linked above, which only occurs if there is already a fetched digest from a provider.  That's a reasonable argument, but I think we also have to consider the `EVP_MD_CTX_copy_ex()` case, though, which is another path to set `ctx->algctx` and seems to allow for further complications.  For example, if we try to make a copy where `out` (already on input to the function) has an algctx set but `in` is legacy, it seems we might be able to create an `EVP_MD_CTX` with non-NULL `algctx` but where `ctx->digest` is a legacy (non-provider) digest, which makes the code in this PR a bit harder to reason about.  (Down in https://github.com/openssl/openssl/blob/6009997abd2594d5a7c0606176f404190922b74d/crypto/evp/digest.c#L566-L568 we only NULL out pointer "in case of error" in the non-legacy case, but logic akin to what you are adding here in this PR in R228 is not present for `EVP_MD_CTX_copy{,_ex}`.)

I would be more comfortable (and able to readily approve a PR) if we were `ossl_assert()`ing that `type->prov` was non-NULL (if `ctx->algctx != NULL`) up around R240 where we have a similar sanity check for `ctx->digest`.  I could try to put together a PR to clean up the legacy case of `EVP_MD_CTX_copy_ex()` if you don't want to try to roll those changes into this PR.

Comment (t8m) [reply crypto/evp/digest.c:251]: I do not see how in the legacy path in EVP_MD_CTX_copy_ex you could end up with non-null algctx even in case of error.

Comment (kaduk) [reply crypto/evp/digest.c:251]: I was envisioning something like
```
EVP_MD_CTX *questionable, *provided_ctx, *legacy_ctx;
EVP_MD *provided, *legacy;
ENGINE *e = ...; /* something not NULL */
provided = EVP_MD_fetch(...);
legacy = EVP_sha256();
provided_ctx = EVP_MD_CTX_new();
legacy_ctx = EVP_MD_CTX_new();
EVP_DigestInit_ex(provided_ctx, provided, NULL);
EVP_DigestInit_ex(legacy_ctx, legacy, e);
EVP_MD_CTX_copy_ex(questionable, provided_ctx);
EVP_MD_CTX_copy_ex(questionable, legacy_ctx);
```
As I understand it, the penultimate line will set `questionable->algctx` and the last line will not clear it.
(I have not tested this.)

Comment (t8m) [reply crypto/evp/digest.c:251]: It will be cleared by https://github.com/openssl/openssl/blob/6009997abd2594d5a7c0606176f404190922b74d/crypto/evp/digest.c#L612

**Comment by t8m** [comment]
ping again for the second review

File: crypto/evp/digest.c
Comment (beldmit) [inline crypto/evp/digest.c:249]: Can't this change be united with a previous one or moved to a function?

Comment (t8m) [reply crypto/evp/digest.c:249]: I can do that.

**Comment by t8m** [comment]
@paulidale still OK?

File: crypto/evp/digest.c
Comment (kaduk) [reply crypto/evp/digest.c:251]: Okay, I see it now.  Thanks for the pointer.
I would still be more comfortable if we asserted that the state of the object is what we expected, but am willing to approve the change without it.  I will see if I can review the update prompted by Dmitry's review.
(Sorry for the long response times here, I'm travelling at the moment.)

**Review by kaduk — APPROVED** [review]
I still think that we should assert that the `type` we're using is a provided type if we are going to reuse the `algctx`.  That is currently true for all possible codepaths, and by writing the assertion we document that when we write this code we are making the assumption that the type is provided.  Having the assumption documented is very useful in the case of any future changes to this code and makes the code more maintainable.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master branch. Thank you for the discussion and reviews.

**Pull request closed without merging**
